helix-view/src/editor.rs

```diff
@@ -103,14 +103,37 @@ pub struct Config {
     /// Whether to display infoboxes. Defaults to true.
     pub auto_info: bool,
     pub file_picker: FilePickerConfig,
+    /// Shape for cursor in each mode
+    pub cursor_shape: CursorShapeConfig,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+#[serde(transparent)]
+pub struct CursorShapeConfig(HashMap<Mode, CursorKind>);
+
+impl std::ops::Deref for CursorShapeConfig {
+    type Target = HashMap<Mode, CursorKind>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for CursorShapeConfig {
+    fn default() -> Self {
+        Self(hashmap!(
+            Mode::Insert => CursorKind::Block,
+            Mode::Normal => CursorKind::Block,
+            Mode::Select => CursorKind::Block,
+        ))
+    }
```

Comment:
We could use a `[u8; 3]` if you assigned indices to `enum Mode`, avoiding hashmap lookups.

Reply:
`Mode` now indexes into a `[Cursorkind; 3]` wrapped inside `CursorShapeConfig`.